File: fractal_sequences/recurrence.py
"""Generic recurrence definitions."""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass

from .models import Number

StepFunction = Callable[[Number, int], Number]


@dataclass(frozen=True)
class Recurrence:
    """Definition of x[n + 1] = f(x[n], n)."""

    name: str
    initial_value: Number
    step: StepFunction

    def next_value(self, current_value: Number, index: int) -> Number:
        return self.step(current_value, index)

File: src/recurrence.rs
use std::sync::Arc;

use crate::models::Number;

pub type StepFunction = Arc<dyn Fn(Number, usize) -> Number + Send + Sync>;

#[derive(Clone)]
pub struct Recurrence {
    pub name: String,
    pub initial_value: Number,
    pub step: StepFunction,
}

impl Recurrence {
    pub fn new(name: impl Into<String>, initial_value: Number, step: StepFunction) -> Self {
        Self {
            name: name.into(),
            initial_value,
            step,
        }
    }

    pub fn next_value(&self, current_value: Number, index: usize) -> Number {
        (self.step)(current_value, index)
    }
}
